### DISCUSSION TITLE
Problems hitting a localhost API

### DISCUSSION THREAD

**Discussion by miguelisidoro** [axios/axios#6722] [Q&A]
Hello,

I am using React with Remix and I am using fetch to call an API that runs in my localhost (developed with C# and .NET 8).
The ports of my React application and web api are different but both run in localhost.

The code runs in a Remix (on server-side, i.e. node.js) and is the following:

```
response = await fetch(apiURL, {
        method: "GET",
        headers: { "Content-Type": "application/json" },
      });
```

I would say that normally there is a problem with the API (a http 500 error) but here the API is not being called.
Any ideias on how to solve this?

If I use axios, I get an error saying "self signed certificate" and a huge stack trace but with no additional details.

The axios code I tried is:

```
 get = async <T>(
    url: string,
    config?: AxiosRequestConfig<any> | undefined
  ) => {
    try {
      const res = await this.api.get<T>(url, {
        ...config,
      });
      return res.data;
    } catch (err) {
      console.log("Error:" + JSON.stringify(err));
      if (axios.isAxiosError(err) && err.response) {
        console.log("Error:" + JSON.stringify(err.response));
        const error = (err.response.data?.detail && {
          cause: {
            status: err.response.status,
            detail: err.response.data?.detail,
          },
        }) || {
          cause: err.response.status,
        };
        throw new Error(
          error.detail || "Something went wrong: " + err.toString()
        );
      }
      throw new Error("Something went wrong");
    }
  };
```

Thanks

**Comment by jasonsaayman** [comment]
This does not look like an Axios-specific problem. The key clue is:

```txt
self-signed certificate
```

That means the request is failing during the HTTPS/TLS handshake before your API route is reached. So your .NET API may appear to be “not being called” because Node rejects the local development certificate before an HTTP request completes.

If this is only local development, the simplest fix is to call the HTTP endpoint instead of HTTPS:

```ts
const apiURL = 'http://localhost:5000/api/your-endpoint';
```

Check the exact ports printed by your .NET app when it starts:

```txt
Now listening on: http://localhost:5000
Now listening on: https://localhost:5001
```

If you need to use HTTPS locally, trust the .NET development certificate:

```sh
dotnet dev-certs https --trust
```

Then restart both the .NET API and the Remix dev server.

For Axios, you can temporarily disable certificate validation in local development only:

```ts
import axios from 'axios';
import https from 'node:https';

const api = axios.create({
  baseURL: 'https://localhost:5001',
  httpsAgent: new https.Agent({
    rejectUnauthorized: false
  })
});
```

Do not use `rejectUnauthorized: false` in production. It disables TLS certificate verification.

Also, since this code runs server-side in Remix, CORS is probably not the issue. CORS applies to browser requests. A Remix loader/action running in Node is making a server-to-server request.

A few other things to check:

- If Remix is running in Docker, `localhost` means the Remix container, not your host machine. Use `host.docker.internal` or the Docker service name.
- Make sure the API URL uses the correct protocol and port: `http://localhost:5000` vs `https://localhost:5001`.
- Try calling the API from the same environment as Remix using `curl`.
- Try `127.0.0.1` instead of `localhost` if there is an IPv4/IPv6 binding issue.
- For a `GET` request, you usually do not need `Content-Type: application/json` because there is no request body.

For debugging Axios errors, avoid `JSON.stringify(err)`. Axios errors contain useful fields that may not serialise well. Log this instead:

```ts
if (axios.isAxiosError(err)) {
  console.log({
    message: err.message,
    code: err.code,
    status: err.response?.status,
    data: err.response?.data,
    url: err.config?.url
  });
}
```

So the likely fix is: use the HTTP localhost URL during development, or make Node trust the local .NET HTTPS certificate.
